haskellPackages.taffybar: fix build

rvl opened the pull request:
###### Motivation for this change

Add patches to allow building with gi-gdkpixbuf_2_0_26

Upstream PRs:
 - https://github.com/taffybar/gtk-sni-tray/pull/25
 - https://github.com/taffybar/taffybar/pull/507

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

rvl committed:
haskellPackages.taffybar: fix build

Add patches to allow building with gi-gdkpixbuf_2_0_26

Upstream PRs:
 - https://github.com/taffybar/gtk-sni-tray/pull/25
 - https://github.com/taffybar/taffybar/pull/507

sternenseemann commented on pkgs/development/haskell-modules/configuration-common.nix:
```diff
@@ -1271,6 +1271,20 @@ self: super: {
 
+  # 2021-05-12: gi-gdkpixbuf_2_0_26 fix
+  # https://github.com/taffybar/gtk-sni-tray/pull/25
+  gtk-sni-tray = markUnbroken (appendPatch super.gtk-sni-tray (pkgs.fetchpatch {
```
Please don't use `markUnbroken` in nixpkgs itself, instead remove the packages in question from `pkgs/development/haskell-modules/configuration-hackage2nix/broken.yaml` and run `./maintainers/scripts/haskell/regenerate-hackage-packages.sh`.

rvl replied:
OK thanks, fixed.

sternenseemann commented:
`taffybar` probably warrants a fix in master, we probably should add that to `mergeable` (we really need to expand the list tbh).

maralorn commented:
> `taffybar` probably warrants a fix in master, we probably should add that to `mergeable` (we really need to expand the list tbh).

Does it? I knew that taffybar was broken. I even opened an issue. But taffybar does not have a maintainer. So I didn‘t care very much. Before we put stuff in mergeable we should find maintainers for those packages.

@rvl do you by any chance want to be a taffybar maintainer?

Cons: You get pinged when it breaks.
Pros: You get pinged when it breaks. And we know that someone cares about the package.

It’s nothing more than that. You are free to ignore the ping when it breaks.

sternenseemann commented:
> Does it? I knew that taffybar was broken. I even opened an issue. But taffybar does not have a maintainer. So I didn‘t care very much. Before we put stuff in mergeable we should find maintainers for those packages.

That is too careless an approach. People use taffybar as part of their desktop setup and as far as I am aware it is not part of any test which controls the `unstable` channel. Thus we risk breaking downstream user's machines via channel updates if we adopt such a carefree well-it-doesn't-have-a-maintainer approach.

I'm more comfortable with breaking development tools and library dependencies, but I feel like we should invest extra time in keeping user-facing derivations working. The channel may be called `unstable`, but that doesn't mean we shouldn't care about breaking stuff.

maralorn commented:
> IMO that is too careless an approach. People use taffybar as part of their desktop setup and as far as I am aware it is not part of any test which controls the `unstable` channel. Thus we risk breaking downstream user's machines via channel updates if we adopt such a carefree well-it-doesn't-have-a-maintainer approach.

Well, I didn‘t know that. That’s my point. I have no clue how many users taffybar has. I have heard about maybe 3. The point about maintained+mergeable jobs is that I don‘t have to make these judgement calls. If taffybar is that essential it should at least have a maintainer.
(I am pretty certain that peti didn‘t watch out for breaking taffybar.)

> I'm more comfortable with breaking development tools and library dependencies, but I feel like we should invest extra time in keeping user-facing derivations working. The channel may be called `unstable`, but that doesn't mean we shouldn't care about breaking stuff.

In general I totally agree.

rvl committed:
taffybar: restore ability to override ghcWithPackages arg

rvl committed:
haskellPackages.taffybar: add rvl to meta.maintainers

Add myself as a "maintainer" of taffybar, and also arbtt and lentil,
for which I have made build fixes in the past.

rvl commented:
Yes Taffybar is a desktop component and it has a few users.
There would probably be more users but it's devillishly hard to build and more often broken than not in `nixos-unstable`.
We got it building for `nixos-20.03` and `nixos-20.09` though, and it would be nice to have a working Taffybar in `nixos-21.05` too.
I added myself to the maintainers list, FWIW, and repaired a change which broke my local config.

sternenseemann commented on pkgs/applications/window-managers/taffybar/default.nix:
```diff
@@ -15,5 +15,13 @@ in stdenv.mkDerivation {
+  passthru.env = taffybarEnv;
+
+  meta = taffybar.meta // {
+    maintainers = with lib.maintainers; [ rvl ];
```
It'd be preferrable if you added yourself in `haskellPackages` already: You can do that by adding an to `package-maintainers` in `pkgs/development/haskell-modules/configuration-hackage2nix/main.yaml` (and rerunning `./maintainers/scripts/haskell/regenerate-haskell-packages.nix`).

rvl replied:
Aha right, simple, fixed now.

maralorn commented:
Yes I remember that it was broken for good parts of the last year. But we can certainly try to increase the quality level here.

sternenseemann commented on pkgs/applications/window-managers/taffybar/default.nix:
```diff
@@ -15,5 +15,11 @@ in stdenv.mkDerivation {
+  # For hacking purposes
+  passthru.env = taffybarEnv;
+
+  # Will be faster to build the wrapper locally then to fetch it from a binary cache.
```
```suggestion
  # Will be faster to build the wrapper locally than to fetch it from a binary cache.
```

sternenseemann commented:
@ofborg build taffybar

maralorn commented on pkgs/applications/window-managers/taffybar/default.nix:
```diff
@@ -15,5 +15,11 @@ in stdenv.mkDerivation {
+  passthru.env = taffybarEnv;
+
+  # Will be faster to build the wrapper locally than to fetch it from a binary cache.
+  preferLocalBuild = true;
```
As far as I can tell `preferLocalBuild` will not prevent fetching from binary caches.
(There is a lot of uncertainty around this topic, I‘d be happy about a definitive answer.)

sternenseemann replied:
`allowSubstitutes` does that. `preferLocalBuild` prevents remote builders from being used as much as possible.

rvl replied:
It's cargo-culted from other wrappers.
It seems like something users would want, given that the packages parameter is configurable.
But I have no idea whether it's effective.
Perhaps it needs `allowSubstitutes = false;` too?

maralorn replied:
We can leave the flag. (It doesn‘t really matter that much either way.) I just believe the comment is wrong.

rvl replied:
OK. I fixed the comment and also added `allowSubstitutes=false` because these two flags seem to be commonly used together for trivial derivations which often need to be built locally.

sternenseemann commented:
@ofborg build taffybar

maralorn commented:
Thank you for fixing this! And for maintaining stuff!

IvanMalison commented:
@maralorn @rvl As the primary maintainer of taffybar (who is also a nix and nixos user) I'd prefer it if I was consulted before things like this were done in nixpkgs. I have merged @rvl 's change to gtk-sni-tray, and bumped its version, and made my own fix for taffybar itself, that should allow us to simply use newer versions that are actually on hackage instead of using patches.

sternenseemann commented:
Feel free to add yourself as a maintainer by adding a corresponding entry to `pkgs/development/haskell-modules/configuration-hackage2nix/main.yml`.

maralorn commented:
Oh, @IvanMalison that is great to know! And yeah, I really think that you should enter yourself as maintainer then.

rvl commented:
Hi @IvanMalison, I did link this PR with the gtk-sni-tray and taffybar PRs. We can simply remove those appendPatch overrides next time the Haskell packages set gets updated - whenever that might be.
